### ISSUE TITLE
[Form] Frustration and miracles

### ISSUE THREAD

**Issue by narcoticfresh** [symfony/symfony#15231]
Hi guys..

The Form Component really is something. I guess you know that. Trying to understand what works how is just an seemingly impossible task; and I'm quite experienced at browsing through codebases.. But man, the Form component.. OMG :octopus:  

**TLDR;**
Below are details, this issue tries to ask
* Is it possible to replace the class `\Symfony\Component\Form\Form`?
* Or: How to easily get *all* *extra data* from all fields of type `form` of a form?
* Or, a related question: How on earth does it work that if i do `$form->add('ss', 'form')` - obviously the `Core\FormType` class gets involved but when i retrieve it later, it is an instance of `\Symfony\Component\Form\Form`? Where does that happen and can this maybe overriden so it uses a different class there?

@webmozart Maybe there is an obvious easy solution to my problem, but I'm not a 'symfony pro' ;-/

**Details**
*The situation*
Imagine a Controller that receives a *n* deep JSON payload. This payload gets decoded and validated through Form. Now, most of this JSON structure is mapped by Models (Doctrine ODM Entities). *But some* sub-properties are just "hashes" - the client is allowed to post whatever he wants there.
Those "hash" subproperties are fields of type `form`, are compound flagged and can have extra fields.

*The problem*
Bottom line our problem is, that all "extra fields" are not returned by `$form->getData()`. We are unable to specify those fields (and their types) as we don't know what will come - so all this data is part of `extraData`.

*The solution?*
So i thought - OK - let's modify the `FormFactory` so it'll return our Custom extension of `Form` - one that also returns `extraData` when `getData` is called. Easy right..? Noo, not easy.. 

My idea let me inspect FormFactory, the FormBuilders, the ResolvedFormTypes, whatnot.. and at all important places (like `formBuilder->getForm()`) - the classes I want to override/replace are instanciated static - no usage of the DIC..(!)

**Comment by Tobion** [comment]
Please use stack overflow for support questions. We use github issues for bugs or concrete feature requests. Thanks.
